### ISSUE TITLE
BufferUnderflowException on restarting node

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#2359]
In the chaos meeting we observed that if we for example restart the leader for a partition the new leader can't take over because in atomix an BufferOverflowException occurs.

```
08:41:09.730 [raft-server-raft-partition-2-state] ERROR io.atomix.protocols.raft.impl.RaftServiceManager - RaftServer{raft-partition-2} - Failed to read snapshot
java.nio.BufferUnderflowException: null
	at io.atomix.storage.buffer.AbstractBuffer.checkSlice(AbstractBuffer.java:348) ~[atomix-storage-3.1.5.jar:?]
	at io.atomix.storage.buffer.AbstractBuffer.slice(AbstractBuffer.java:178) ~[atomix-storage-3.1.5.jar:?]
	at io.atomix.protocols.raft.impl.RaftServiceManager.install(RaftServiceManager.java:501) ~[atomix-raft-3.1.5.jar:?]
	at io.atomix.protocols.raft.impl.RaftServiceManager.lambda$apply$12(RaftServiceManager.java:419) ~[atomix-raft-3.1.5.jar:?]
	at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.1.5.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_202]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_202]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_202]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_202]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_202]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_202]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
```

Might be related to https://github.com/atomix/atomix/issues/1007
 and see https://github.com/zeebe-io/zeebe/issues/1874#issuecomment-458029714

**Comment by npepinpe** [comment]
I'm putting this on hold as this seems to be a bug in Atomix and I'd rather focus on fixing bugs in our own codebase, but here's what I gathered:

Both this and #2435 seem to be related to corrupt snapshots, although both have slightly different causes:

When Atomix takes a snapshot, it reserves the first 64 bytes for a `SnapshotDescriptor` which contains the snapshot index, timestamp, etc., and a `lock` property which allows it determine if the snapshot was completed or not (by setting that byte to 1 as the very last step of writing the snapshot, and refusing to write anything more if that byte is already 1). After those 64 bytes, it will reserve 4 bytes for the length of the snapshot data, write the data, then go back to offset 64 and write the length of the data.

So in this case, here, what happened is it wrote some length X which was greater than the actual snapshot file length (assuming there is no bug in the way it allocates the file buffer, of course, and the length was properly calculated). 

In the case of #2435 it's worse in the sense that it reads the descriptor (maybe? it's possible even the descriptor was mangled as it only checks if byte at offset 20 is equal to 1), sees the snapshot was 'locked', but when it tries to read its length from offset 64 it fails as there are not enough bytes to read.

So both cases seem to point to snapshot corruption, but reading the code it seems unlikely, and it is very hard to reproduce this. It's also hard to pin point if it gets corrupted locally or when its replicated over the network. Replication works by buffering the whole snapshot in memory before flushing to disk as a single write, so my guess is it's most likely local?

**Comment by npepinpe** [comment]
After #2707 this is greatly mitigated; will happen but the impact is not as bad as before. Should we leave this open (as it's not solved) or reopen it when it happens again? @menski

**Comment by npepinpe** [comment]
Closed by https://github.com/zeebe-io/atomix/pull/21
